### PR TITLE
Checking for transaction in Realm.refresh and throwing if inside of a Transaction

### PR DESCRIPTION
**Pull request by donnfelker**
... as the Realm is always kept up to date.

This PR is for #1712

### CODE REVIEW

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 1]
```diff
@@ -56,6 +56,7 @@
     private static final String INCORRECT_THREAD_MESSAGE = "Realm access from incorrect thread. Realm objects can only be accessed on the thread they were created.";
     private static final String CLOSED_REALM_MESSAGE = "This Realm instance has already been closed, making it unusable.";
     private static final String DIFFERENT_KEY_MESSAGE = "Wrong key used to decrypt Realm.";
+    private static final String CANNOT_REFRESH_INSIDE_OF_TRANSACTION_MESSAGE = "Cannot refresh inside of a transaction, the Realm is always kept up to date.";
```
Comment (emanuelez) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: Realm are only kept up to date if they are in a Looper thread so I would truncate the message at the comma.

**Comment by emanuelez** [comment]
:+1: apart from the minor comment

File: realm/realm-library/src/androidTest/java/io/realm/RealmTest.java [hunk 2]
```diff
@@ -2339,6 +2339,18 @@ public void run() {
         }
     }
 
+    public void testCannotRefreshInTransaction() {
+        testRealm.beginTransaction();
+        try {
+            testRealm.refresh();
+            fail("Cannot refresh in a Transaction. Expected IllegalStateException to be thrown.");
+        } catch (IllegalStateException ex) {
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTest.java]: Nitpick: normally we use `e` or `ignored` for this.

File: realm/realm-library/src/main/java/io/realm/BaseRealm.java [hunk 3]
```diff
@@ -281,6 +282,10 @@ public void writeEncryptedCopyTo(File destination, byte[] key) throws java.io.IO
     @SuppressWarnings("UnusedDeclaration")
     public void refresh() {
         checkIfValid();
+        // Check if we are already in a transaction
```
Comment (cmelchior) [inline realm/realm-library/src/main/java/io/realm/BaseRealm.java]: This comment doesn't say anything the if statement doesn't. I would remove it.

**Comment by cmelchior** [comment]
Apart from a few nitpicks it looks good. We need to update the `changelog` though

**Comment by donnfelker** [comment]
Ok, all fixed up.

**Comment by cmelchior** [comment]
:+1:

File: realm/realm-library/src/androidTest/java/io/realm/RealmTest.java [hunk 4]
```diff
@@ -2339,6 +2339,18 @@ public void run() {
...
 
+    public void testCannotRefreshInTransaction() {
+        testRealm.beginTransaction();
+        try {
+            testRealm.refresh();
+            fail("Cannot refresh in a Transaction. Expected IllegalStateException to be thrown.");
+        } catch (IllegalStateException e) {
+            assertEquals(e.getMessage(), "Cannot refresh inside of a transaction.");
+        } finally {
+          testRealm.cancelTransaction();
```
Comment (cmelchior) [inline realm/realm-library/src/androidTest/java/io/realm/RealmTest.java]: Indentation looks a bit funny here?

**Comment by cmelchior** [comment]
@donnfelker This looks ready to merge as well as soon as it has been squashed.

**Comment by donnfelker** [comment]
@cmelchior Squashed and ready to go.

File: changelog.txt [hunk 5]
```diff
@@ -2,6 +2,7 @@
  * Added Realm.isEmpty().
  * RealmQuery.isNull() and RealmQuery.isNotNull() now throw IllegalArgumentException instead of RealmError if the fieldname is a linked field and the last element is a link (#1693).
  * Setters in managed object for RealmObject and RealmList now throw IllegalArgumentException if the value contains an invalid (standalone, removed, closed, from different Realm) object (#1749).
+ * Attempting to refresh a Realm while a transaction is in process will now throw an IllegalArgumentException (#1712)
```
Comment (bmunkholm) [inline changelog.txt]: IllegalArgumentException -> IllegalStateException

**Comment by donnfelker** [comment]
@bmunkholm fixed

**Comment by donnfelker** [comment]
Good catch btw! :)

**Comment by cmelchior** [comment]
Great :+1: Merged

**Pull request merged**